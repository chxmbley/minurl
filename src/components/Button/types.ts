import type { HTMLProps } from 'react';

/** Props for the Button component */
export type ButtonProps = HTMLProps<HTMLButtonElement> & {
  /** Type of the button. Note that this type is explicitly defined because HTMLProps<HTMLButtonElement> defines the `type` prop incorrectly. */
  type?: 'button' | 'submit' | 'reset';
};
