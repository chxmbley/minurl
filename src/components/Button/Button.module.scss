@mixin disabled-colors {
  @apply bg-slate-200 dark:bg-slate-800 text-slate-400 dark:text-slate-600;
}

.button {
  @apply flex justify-center items-center text-slate-600 dark:text-slate-100 transition-colors outline-none rounded px-3 h-8 text-xs uppercase font-medium tracking-wider;

  &:hover,
  &:focus {
    @apply bg-slate-200 dark:bg-slate-800;
  }

  &:disabled {
    @include disabled-colors;
    @apply cursor-not-allowed;
  }
}

.button.primary {
  @apply bg-purple-600 text-white;

  &:hover,
  &:focus {
    @apply bg-purple-500;
  }

  &:disabled {
    @include disabled-colors;
  }
}

.button.accent {
  @apply bg-cyan-500 text-white;

  &:hover,
  &:focus {
    @apply bg-cyan-400;
  }

  &:disabled {
    @include disabled-colors;
  }
}

.button.round {
  @apply aspect-square rounded-full;
}
