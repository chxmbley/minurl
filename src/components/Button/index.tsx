import cn from 'classnames';
import type { FC } from 'react';
import type { ButtonProps } from './types';

/** Styled Button component */
const Button: FC<ButtonProps> = ({ className, ...props }) => (
  <button
    className={cn(
      className,
      'flex justify-center items-center bg-cyan-500 dark:bg-cyan-500 text-white transition-colors hover:bg-cyan-400 focus:bg-cyan-400 outline-none disabled:bg-slate-300 disabled:text-slate-400 disabled:dark:bg-slate-700 disabled:dark:text-slate-500 disabled:cursor-not-allowed',
    )}
    {...props}
  />
);

export default Button;
