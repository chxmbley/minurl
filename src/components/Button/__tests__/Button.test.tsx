import { render } from '@testing-library/react';
import Button from '~components/Button';

describe('Button', () => {
  it('renders a Button component', () => {
    const container = render(<Button />);
    expect(container).toMatchSnapshot();
  });

  it('renders a Button component with text children', () => {
    const container = render(<Button>Hello world!</Button>);
    expect(container).toMatchSnapshot();
  });

  it('renders a Button component with React node children', () => {
    const container = render(
      <Button>
        Hello <span className="text-red-500">world!</span>
      </Button>,
    );

    expect(container).toMatchSnapshot();
  });

  it('adds a provided className to the component', () => {
    const customClassName = 'my-custom-classname';
    const container = render(<Button data-testid="button" className={customClassName} />);
    expect(container).toMatchSnapshot();
    expect(container.getByTestId('button')).toHaveClass(customClassName);
  });
});
