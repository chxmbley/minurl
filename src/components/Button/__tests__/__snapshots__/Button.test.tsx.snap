// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button adds a provided className to the component 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <button
        class="my-custom-classname flex justify-center items-center bg-cyan-500 dark:bg-cyan-500 text-white transition-colors hover:bg-cyan-400 focus:bg-cyan-400 outline-none disabled:bg-slate-300 disabled:text-slate-400 disabled:dark:bg-slate-700 disabled:dark:text-slate-500 disabled:cursor-not-allowed"
        data-testid="button"
      />
    </div>
  </body>,
  "container": <div>
    <button
      class="my-custom-classname flex justify-center items-center bg-cyan-500 dark:bg-cyan-500 text-white transition-colors hover:bg-cyan-400 focus:bg-cyan-400 outline-none disabled:bg-slate-300 disabled:text-slate-400 disabled:dark:bg-slate-700 disabled:dark:text-slate-500 disabled:cursor-not-allowed"
      data-testid="button"
    />
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Button renders a Button component 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <button
        class="flex justify-center items-center bg-cyan-500 dark:bg-cyan-500 text-white transition-colors hover:bg-cyan-400 focus:bg-cyan-400 outline-none disabled:bg-slate-300 disabled:text-slate-400 disabled:dark:bg-slate-700 disabled:dark:text-slate-500 disabled:cursor-not-allowed"
      />
    </div>
  </body>,
  "container": <div>
    <button
      class="flex justify-center items-center bg-cyan-500 dark:bg-cyan-500 text-white transition-colors hover:bg-cyan-400 focus:bg-cyan-400 outline-none disabled:bg-slate-300 disabled:text-slate-400 disabled:dark:bg-slate-700 disabled:dark:text-slate-500 disabled:cursor-not-allowed"
    />
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Button renders a Button component with React node children 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <button
        class="flex justify-center items-center bg-cyan-500 dark:bg-cyan-500 text-white transition-colors hover:bg-cyan-400 focus:bg-cyan-400 outline-none disabled:bg-slate-300 disabled:text-slate-400 disabled:dark:bg-slate-700 disabled:dark:text-slate-500 disabled:cursor-not-allowed"
      >
        Hello 
        <span
          class="text-red-500"
        >
          world!
        </span>
      </button>
    </div>
  </body>,
  "container": <div>
    <button
      class="flex justify-center items-center bg-cyan-500 dark:bg-cyan-500 text-white transition-colors hover:bg-cyan-400 focus:bg-cyan-400 outline-none disabled:bg-slate-300 disabled:text-slate-400 disabled:dark:bg-slate-700 disabled:dark:text-slate-500 disabled:cursor-not-allowed"
    >
      Hello 
      <span
        class="text-red-500"
      >
        world!
      </span>
    </button>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Button renders a Button component with text children 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <button
        class="flex justify-center items-center bg-cyan-500 dark:bg-cyan-500 text-white transition-colors hover:bg-cyan-400 focus:bg-cyan-400 outline-none disabled:bg-slate-300 disabled:text-slate-400 disabled:dark:bg-slate-700 disabled:dark:text-slate-500 disabled:cursor-not-allowed"
      >
        Hello world!
      </button>
    </div>
  </body>,
  "container": <div>
    <button
      class="flex justify-center items-center bg-cyan-500 dark:bg-cyan-500 text-white transition-colors hover:bg-cyan-400 focus:bg-cyan-400 outline-none disabled:bg-slate-300 disabled:text-slate-400 disabled:dark:bg-slate-700 disabled:dark:text-slate-500 disabled:cursor-not-allowed"
    >
      Hello world!
    </button>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
